总的来说，参考TCGA的格式,这里不考虑DEPMAP

# 表达
exp_df: 第一列是`Hugo_Symbol`，后面每一列是样本的表达值，是表达数据的主要交互格式  
exp_mat: 由exp_df生成，gene*sample  

# 突变
MAF格式：主要有这些列
`Hugo_Symbol,Chromosome,Start_position,End_position,Strand,Variant_Classification,Variant_Type,Reference_Allele,Tumor_Seq_Allele1,Tumor_Seq_Allele2,Tumor_Sample_Barcode,AAChange,ITDorNOT`  
外加一列`sample`,用于与其他数据交互

# cnv
待补全

# 药敏
`inhibitor,sample,ic50,auc`


# 生存
`sample,OS,OSS`





---------------------------

# ccle数据


CCLE_drug_family.tsv             


CRISPR_gene_effect_cleaned.tsv: sample*gene (sample)
CRISPR_gene_effect_cleaned_2.tsv: gene*sample (Hugo_Symbol)
CRISPR_gene_effect_cleaned_raw.tsv: sample*gene (sample)

RNAi_gene_effect_cleaned.tsv: gene*sample (Hugo_Symbol)

CCLE_expression_cleaned.tsv      

CCLE_GDSC1_drug_sen_cleaned.tsv  
CCLE_GDSC2_drug_sen_cleaned.tsv

sample_info.tsv

CCLE_mutations_cleaned.tsv
```
Hugo_Symbol,Entrez_Gene_Id,NCBI_Build,Chromosome,Start_position,End_position,Strand,Variant_Classification,Variant_Type,Reference_Allele,Tumor_Seq_Allele1,dbSNP_RS,dbSNP_Val_Status,Genome_Change,Annotation_Transcript,DepMap_ID,cDNA_Change,Codon_Change,Protein_Change,isDeleterious,isTCGAhotspot,TCGAhsCnt,isCOSMIChotspot,COSMIChsCnt,ExAC_AF,Variant_annotation,CGA_WES_AC,HC_AC,RD_AC,RNAseq_AC,SangerWES_AC,WGS_AC,Tumor_Sample_Barcode,Tumor_Seq_Allele2,sample
```

数据处理工具






